executor/builder.go

```diff
@@ -2040,7 +2040,65 @@ func (b *executorBuilder) buildAnalyzeIndexIncremental(task plannercore.AnalyzeI
 	return analyzeTask
 }
 
+func (b *executorBuilder) buildAnalyzeSamplingPushdown(
+	task plannercore.AnalyzeColumnsTask,
+	opts map[ast.AnalyzeOptionType]uint64,
+	autoAnalyze string,
+) *analyzeTask {
+	availableIdx := make([]*model.IndexInfo, 0, len(task.Indexes))
+	colGroups := make([]*tipb.AnalyzeColumnGroup, 0, len(task.Indexes))
+	if len(task.Indexes) > 0 {
+		for _, idx := range task.Indexes {
+			availableIdx = append(availableIdx, idx)
+			colGroup := &tipb.AnalyzeColumnGroup{
+				ColumnOffsets: make([]int64, 0, len(idx.Columns)),
+			}
+			for _, col := range idx.Columns {
+				colGroup.ColumnOffsets = append(colGroup.ColumnOffsets, int64(col.Offset))
+			}
+			colGroups = append(colGroups, colGroup)
+		}
+	}
+
+	_, offset := timeutil.Zone(b.ctx.GetSessionVars().Location())
+	sc := b.ctx.GetSessionVars().StmtCtx
+	e := &AnalyzeColumnsExec{
+		ctx:         b.ctx,
+		tableID:     task.TableID,
+		colsInfo:    task.ColsInfo,
+		handleCols:  task.HandleCols,
+		concurrency: b.ctx.GetSessionVars().DistSQLScanConcurrency(),
+		analyzePB: &tipb.AnalyzeReq{
+			Tp:             tipb.AnalyzeType_TypeFullSampling,
+			Flags:          sc.PushDownFlags(),
+			TimeZoneOffset: offset,
+		},
+		opts:       opts,
+		analyzeVer: task.StatsVersion,
+		indexes:    availableIdx,
+	}
+	e.analyzePB.ColReq = &tipb.AnalyzeColumnsReq{
+		BucketSize:   int64(opts[ast.AnalyzeOptNumBuckets]),
+		SampleSize:   int64(opts[ast.AnalyzeOptNumSamples]),
+		SketchSize:   maxSketchSize,
+		ColumnsInfo:  util.ColumnsToProto(task.ColsInfo, task.TblInfo.PKIsHandle),
+		ColumnGroups: colGroups,
+	}
+	if task.TblInfo != nil {
+		e.analyzePB.ColReq.PrimaryColumnIds = tables.TryGetCommonPkColumnIds(task.TblInfo)
+		if task.TblInfo.IsCommonHandle {
+			e.analyzePB.ColReq.PrimaryPrefixColumnIds = tables.PrimaryPrefixColumnIDs(task.TblInfo)
```

Comment:
How is the `PrimaryPrefixColumnIds` used in coprocessor? why do we need to record it?

Reply:
😂 It's copied from original. Seems not used, let me check it again.

Reply:
It's used in unistore to decode row value.